compiler/rustc_session/src/session.rs

```diff
@@ -449,15 +449,29 @@ impl Session {
         )
     }
 
-    /// Returns a list of directories where target-specific tool binaries are located.
+    /// Returns a list of directories where target-specific tool binaries are located. Some fallback
+    /// directories are also returned, for example if `--sysroot` is used but tools are missing
+    /// (#125246): we also add the bin directories to the sysroot where rustc is located.
     pub fn get_tools_search_paths(&self, self_contained: bool) -> Vec<PathBuf> {
         let rustlib_path = rustc_target::target_rustlib_path(&self.sysroot, config::host_triple());
         let p = PathBuf::from_iter([
             Path::new(&self.sysroot),
             Path::new(&rustlib_path),
             Path::new("bin"),
         ]);
-        if self_contained { vec![p.clone(), p.join("self-contained")] } else { vec![p] }
+
+        let fallback_sysroot_paths = filesearch::sysroot_candidates()
+            .into_iter()
+            .map(|sysroot| filesearch::make_target_bin_path(&sysroot, config::host_triple()));
+
+        if self_contained {
+            [p.clone(), p.join("self-contained")]
+                .into_iter()
+                .chain(fallback_sysroot_paths)
```

Comment:
Hmm, I think the `self-contained` subdirectory should be added to all higher-level directories, including those from the fallback sysroot.

Reply:
There probably won't be any tools in these folders now, but at least it'll work when we move everything there.

Reply:
Done. I also used the new helper as a slight cleanup I had forgotten to do.